### ISSUE TITLE
tropical/einmax1 gradients are silently wrong for overlapping windows (last-write-wins gate)

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#512]
## Symptom

The gradients of `tropical` (`@^+`) and `einmax1` (`@^^`) are exact only when the reduction windows don't overlap (stride ≥ window). With overlapping windows the gradient is silently wrong in both directions — positions can receive gradient from windows they did not win, and can lose gradient from windows they did win. Nothing in `Operation.tropical`'s docstring mentions this (it documents only the separate g2 index-space limitation), and all existing tests (`test_tropical_kernel.ml`, `test_max_pool2d.ml`, `clamped_window_pool.ml`) use stride = window, so the suite never exercises it. `max_pool2d` with `stride < window_size` (AlexNet-style overlapping pooling) trains on wrong gradients today.

## Minimal repro (verified on current master)

1-D max-slide, window 2, stride 1, zero kernel, `sum` loss:

```ocaml
let%op input = { x = [ 5.; 3.; 1.; 2.; 5. ] } in
let%op output = Shape.set_dim k win; input @^+ "o< + k,; k, => o," [ "k" ] (0.0 + 0.0) in
let%op loss = output ++ "...|... => |->0" in
```

Forward is correct: output `[5;3;2;5]`, loss 15. Argmaxes are i=0 (window 0), i=1 (window 1), i=3 (window 2), i=4 (window 3), so the exact input gradient is `[1;1;0;1;1]`. Observed: `[1;2;0;0;1]` — i=1 over-credited (it also collects the gradient of window 0, which i=0 won), i=3's gradient lost (window 3, the last to write i=3's gate bit, saw 5 ≠ 2).

## Root cause

The gradient gate is materialized as a tensor over the *input* index space:

```ocaml
{ sum_rhs1 } =:@^ add (t1, t2);
{ cond_rhs1 } =: eq (t, sum_rhs1);
g1 =+ where cond_rhs1 g 0;
```

`cond_rhs1[i]` is one bit per input position, but the ground truth is per (window, position) pair: with overlap, several windows write conflicting bits through the non-accumulating `=:` and the last write wins; the subsequent `=+` then applies that single bit to ALL windows touching `i`.

This shape choice is forced, not accidental: `Assignments` is a flat IR (one accumulation + one primitive op per statement, tensor-node operands only — a nested `g1 =+ where (eq (v, add (v1, v2))) g 0` is rejected with "insufficient slot filler information"), and the `%cd` inline-declaration suffix vocabulary can only mint intermediates shaped like an existing operand slot (`_rhs1`, `_lhs`, ...). No slot indexes the (output × kernel) pair space that an exact mask needs.

## Is this inherent? No — other frameworks are exact here

- PyTorch `max_pool2d` saves argmax indices in the forward pass and scatters gradients to them; overlapping pooling is exact (single-winner tie semantics).
- XLA/JAX handle `reduce_window(max)` gradients with a *dedicated* `SelectAndScatter` op — evidence that this computation genuinely does not decompose into flat einsum-style assignments, which is exactly where OCANNL's IR pinches.
- Tie handling differs legitimately across frameworks (PyTorch max_pool: single winner; `torch.amax`/JAX: split among ties). The current OCANNL gate splits among ties in the non-overlapping case; that part is defensible and separate from the overlap bug.

## Fix options

1. **Argmax-record route (PyTorch-style, most tractable).** In `op_asn`, after computing `t`, a second last-write-wins pass records the winning position per *output* cell: `argwin[o] = flat index of (i,k) achieving t[o]` — last-write-wins is now per-o, where it is single-writer and therefore correct (and yields deterministic single-winner ties). Backward scatters `g[o]` through the recorded index with dynamic indexing. The Get_dynamic/Set_dynamic machinery from the embedding gather/scatter work (gh-343, gh-466) is the natural substrate; known cost: a dynamic-indexed write currently bails parallelization on the scatter.
2. **Pair-space gate tensor.** Materialize `cond[o,k] = (t[o] == t1[i(o,k)] + t2[k])` over the (output × kernel) product space, then `g1 =+ where cond g 0` with the pair projection. Exact, keeps split-tie semantics, stays in the flat IR — but requires new projection/slot plumbing (a pair-space intermediate has no `%cd` suffix today); essentially a scoped-down SelectAndScatter.
3. **New gated-accumulation primitive.** A ternop `(a == b ? c : 0)` fused into the accumulation fixes `einmax1` (operands v, v1, g) but `tropical` needs four operands (v, v1, v2, g) — doesn't generalize without quaternary ops.
4. **Guard in the meantime (cheap, defuses the footgun).** `tropical`/`einmax1` can detect overlap from the spec (a conv index with stride < window, or more generally a non-injective rhs1 transpose projection) and fail fast or warn, and the docstring should state the limitation. Silent wrongness is the actual footgun; an error converts it into a visible limitation.

Option 4 is worth doing immediately; option 1 looks like the right permanent fix, with option 2 as the more principled but heavier alternative.
